### ISSUE TITLE
Bug: no-unused-expression doesn't report on side-effect free calls?

### ISSUE THREAD

**Issue by kkmuffme** [eslint/eslint#18634]
### Environment

Playground eslint 9.6.0

### What parser are you using?

Default (Espree)

### What did you do?

```js
/* eslint no-unused-expressions: "error", no-unused-vars: "error" */

const a = 15;
const b = 20;

Math.floor(a/b); // should be no-unused-expressions

function foo() { 
  return 2; 
}
foo(); // should be no-unused-expressions

/**
 * @pure
 */
function bar() { 
  return 2; 
}
bar(); // should be no-unused-expressions
```

### What did you expect to happen?

https://eslint.org/docs/latest/rules/no-unused-expressions
>An unused expression which has **no effect on the state of the program** indicates a logic error.

According to this, I'd expect no-unused-expression errors at least for line 6 and 11.

Ideally also @pure jsdoc annotation would be supported to handle cases where eslint isn't able to infer if the function is pure or not and should therefore report no-unused-expressions and give an error in line 19 too

### What actually happened?

No error

### Link to Minimal Reproducible Example

https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXVudXNlZC1leHByZXNzaW9uczogXCJlcnJvclwiLCBuby11bnVzZWQtdmFyczogXCJlcnJvclwiICovXG5cbmNvbnN0IGEgPSAxNTtcbmNvbnN0IGIgPSAyMDtcblxuTWF0aC5mbG9vcihhL2IpOyAvLyBzaG91bGQgYmUgbm8tdW51c2VkLWV4cHJlc3Npb25zXG5cbmZ1bmN0aW9uIGZvbygpIHsgXG4gIHJldHVybiAyOyBcbn1cbmZvbygpOyAvLyBzaG91bGQgYmUgbm8tdW51c2VkLWV4cHJlc3Npb25zXG5cbi8qKlxuICogQHB1cmVcbiAqL1xuZnVuY3Rpb24gYmFyKCkgeyBcbiAgcmV0dXJuIDI7IFxufVxuYmFyKCk7IC8vIHNob3VsZCBiZSBuby11bnVzZWQtZXhwcmVzc2lvbnNcbiIsIm9wdGlvbnMiOnsicnVsZXMiOnt9LCJsYW5ndWFnZU9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9fX19fQ==

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by mdjermanovic** [comment]
Hi @kkmuffme, thanks for the issue!

As stated in [Rule Details](https://eslint.org/docs/latest/rules/no-unused-expressions#rule-details), this rule does not apply to function calls:

> This rule does not apply to function calls or constructor calls with the `new` operator, because they could have *side effects* on the state of the program.

Closing as this works as intended.

**Comment by mdjermanovic** [comment]
The wording may be confusing. The intent was to say that the rule doesn't apply to function calls like `foo()` or constructor calls like `new foo()`. If the sentence sounds confusing, feel free to submit a PR to clarify this.
